numpyro/distributions/continuous.py

```diff
@@ -1415,61 +1378,24 @@ def tree_unflatten(cls, aux_data, params):
         return d
 
 
-class _BaseTruncatedNormal(Distribution):
-    # NB: this is a truncated normal with low=0, scale=1
-    arg_constraints = {"base_loc": constraints.real}
-    reparametrized_params = ["base_loc"]
-    support = constraints.positive
-
-    def __init__(self, base_loc):
-        self.base_loc = base_loc
-        self._normal = Normal(base_loc, 1.)
-        super(_BaseTruncatedNormal, self).__init__(batch_shape=jnp.shape(base_loc))
-
-    def sample(self, key, sample_shape=()):
-        assert is_prng_key(key)
-        size = sample_shape + self.batch_shape
-        # We use inverse transform method:
-        # z ~ icdf(U), where U ~ Uniform(0, 1).
-        u = random.uniform(key, shape=size)
-        # Ref: https://en.wikipedia.org/wiki/Truncated_normal_distribution#Simulating
-        # icdf[cdf_a + u * (1 - cdf_a)] = icdf[1 - (1 - cdf_a)(1 - u)]
-        #                                 = - icdf[(1 - cdf_a)(1 - u)]
-        return self.base_loc - ndtri(ndtr(self.base_loc) * (1 - u))
-
-    @validate_sample
-    def log_prob(self, value):
-        # log(cdf(high) - cdf(low)) = log(1 - cdf(low)) = log(cdf(-low))
-        return self._normal.log_prob(value) - log_ndtr(self.base_loc)
-
-
-class TruncatedNormal(TransformedDistribution):
+class TruncatedNormal(LeftTruncatedDistribution):
```

Comment:
Hmm, I would expect `TruncatedNormal` to be a dispatcher (and similar for `TruncatedCauchy`):
```py
def TruncatedNormal(loc, scale, low=None, high=None, validate_args=None):
    if low is None:
        if high is None:
            return Normal(loc, scale, validate_args=validate_args)
        else:
            return RightTruncatedNormal(loc, scale, high, validate_args=validate_args)
    else:
        if high is None:
            return LeftTruncatedNormal(loc, scale, low, validate_args=validate_args)
        else:
            return BiTruncatedNormal(loc, scale, low, high, validate_args=validate_args)
```
I guess that's out of scope for this PR though...

Reply:
At first, I made a dispatcher and added a deprecation warning message. But some tests require the old interface (e.g. to test mean/var), so I would like to defer doing so in future PRs, if needed.

> revisit the interface to allow two-sided truncated distributions

What is a good way to do this without breaking existing code? I am not sure why I decided to use `low, loc, scale` order back then... Probably we can add a deprecation/warning message in this version, and change the API in the next version.

Reply:
Future PRs sounds good :+1:

> What is a good way to do this without breaking existing code?

Well I'm unsure how to avoid _breaking_ code, but you can avoid _silently_ breaking by requiring kwargs rather than args:
```py
def TruncatedNormal(*, loc=None, scale=None, low=None, high=None, validate_args=None):
    ...
```

Reply:
Haha, nice trick!!! I'll keep this in mind. Currently, I'm not sure if it is better to remove `TruncatedNormal`, `TruncatedCauchy` apis (in favor of `TruncatedDistribution` API), so I'll defer the change to after the next release.